Upgrade Ingress NGINX Controller to version 1.1.1 (#21)
projectsyn/component-ingress-nginx: docs/modules/ROOT/pages/how-tos/upgrade-1.x-to-2.x.adoc

= Upgrade from 1.x to 2.x

This guide describes the steps to perform an upgrade of the component from version 1.x to 2.x.

[WARNING]
====
Version 2.x does use Ingress NGINX Controller >1.0.

Before upgrading to version 2.x, please ensure that your Kubernetes cluster is running the version https://github.com/kubernetes/ingress-nginx#support-versions-table[1.19 or later].
====

====
Requirements

* `commodore`
* `kubectl`
* `shell`
====

== Introduction

The component version 1.x did use Ingress NGINX Controller v0.x and component version 2.x is now using Ingress NGINX Controller v1.x.

> The Ingress NGINX Controller v1.x now requires the use of a `IngressClass`.

Reviewer: ```suggestion
The Ingress NGINX Controller v1.x now requires the use of an `IngressClass`.
```

With the upgrade the component is automatically deploying an `IngressClass` called `nginx` for the controller:

[source,yaml]
----
---
apiVersion: networking.k8s.io/v1
kind: IngressClass
metadata:
  labels:
    app.kubernetes.io/component: controller
  name: nginx
spec:
  controller: k8s.io/ingress-nginx
----

== Step-by-step guide

. If the Ingress NGINX Controller should be the default IngressController, please add this to the configuration:
+
[source,yaml]
----
parameters:
  ingress_nginx:
    helm_values:
      controller:
        ingressClassResource:
          default: true
----
+

> . If not all Ingress resources have a configured `.spec.ingressClassName` configured, please set `watchIngressWithoutClass: true` to still watch for such objects:

Reviewer: ```suggestion
. If not all Ingress resources have a `.spec.ingressClassName` configured, you must set `watchIngressWithoutClass: true` to still watch for such objects:
```

+
[source,yaml]
----
parameters:
  ingress_nginx:
    helm_values:
      controller:
        watchIngressWithoutClass: true
----

== Post upgrade

Since NGINX Controller v1.0 the `.spec.ingressClassName` field should be set in the https://kubernetes.io/docs/concepts/services-networking/ingress/#the-ingress-resource[Ingress resource].
If the `.spec.ingressClassName` is omitted, a default Ingress class should be defined.
If no customisation is done, the `IngressClass` is named `nginx`.

ArgoCD does ignore `.spec.ingressClassName` if not set in the original manifest.

For more information read the NGINX Ingress Controller documentation on the topic https://kubernetes.github.io/ingress-nginx/#i-have-only-one-ingress-controller-in-my-cluster-what-should-i-do[I have only one ingress controller in my cluster. What should I do?].

Existing Ingress objects don't get the default IngressClass.
They need to be recreated or patched.

Patch one Ingress resource:
[source,bash]
----
kubectl patch ingress <ingressname> --patch '{"spec": {"ingressClassName": "nginx"}}' --namespace <namespace>
----

> Patch all Ingress resources on a cluster (BE CAREFULLY WITH IT, IF YOU HAVE MULTIPLE INGRESS CONTROLLERS!):

Reviewer: ```suggestion
Patch all Ingress resources on a cluster (BE CAREFUL WITH IT, IF YOU HAVE MULTIPLE INGRESS CONTROLLERS!):
```

> . Show all Ingress resources who will be patched:

Reviewer: ```suggestion
. Show all Ingress resources that will be patched:
```
